proposals/scopes.md

```diff
@@ -196,106 +196,179 @@ interface OriginalPosition {
 
-  * "originalScopes" is an array of original scope tree descriptors (a string). Each element in the array describes the scope tree of the corresponding "sources" entry.
-  * "generatedRanges" is a single generated range tree descriptor (a string) of the generated file.
+The format of "scopes" is presented in an EBNF-like grammar, with:
 
-Like the "mappings" field, the data in a "generated range tree descriptor" is grouped by line and lines are separated by `;`. Within a line, items are separated by `,`. A "original scope tree descriptor" is NOT grouped by line, but items are separated by `,`.
+  * Three terminals: Signed, unsigned VLQ and comma '`,`'. VLQ terminals are labelled and we denote them with UPPERCASE.
+    We prefix the terminal with `u` or `s` to signify an unsigned or signed VLQ respectively. E.g. the terminal `uLINE` signifies
+    an unsigned VLQ labelled `LINE`.
 
-There are two different kinds of items that will appear in a "original scope descriptor": "Start Original Scope" and "End Original Scope".
-There are two different kinds of items that will appear in the "generated range descriptor": "Start Generated Range" and "End Generated Range".
+  * Non-terminals are denoted with `snake_case`.
 
-The kind of an item can be determined by looking at how many VLQ-encoded numbers it contains: "End" items contain one number, "Start" items contain two or more numbers.
+  * `symbol?` means zero or one `symbol`.
+  * `symbol+` means one or more `symbol`.
 
-Note: Each DATA represents one VLQ number.
+The start symbol is `scopes`:
 
-#### Start Original Scope
+```
+scopes := original_scope_list generated_range_list
+
+original_scope_list :=
+    original_scope_tree
+  | ε
+  | original_scope_list ',' original_scope_tree
+  | original_scope_list ',' ε
+
+generated_range_list :=
+    generated_range_tree
+  | generated_range_list ',' generated_range_tree
+```
+
+Only one top-level `original_scope_tree` per `sources` file is allowed. The n-th top-level `original_scope_tree` describes the scope tree for `sources[n]`. To signify that a certain `sources[m]` authored file doesn't have scopes information available, an empty item must be used.
+Multiple top-level `generated_range_tree`s are allowed, this is especially useful when multiple bundles are straight-up concatenated.
+
+#### Original Scope Trees
+
+```
+original_scope_tree :=
+  original_scope_start
+  original_scope_variables?
+  original_scope_source_index?
+  original_scope_tree?
+  original_scope_end
+```
+
+A scope is delineated by a `original_scope_start` and a `original_scope_end` item. The `original_scope_variables` and `original_scope_source_index`
+item always describe the immediately "surrounding" start/end pair.
+
+```
+original_scope_start :=
+  'B'     // Tag: 0x1 unsigned
+  uFLAGS
+  uLINE
+  uCOLUMN
+  sNAME?  // Present if FLAGS<0> is set.
+  sKIND?  // Present if FLAGS<1> is set.
+
+original_scope_variables :=
+  'D'        // Tag: 0x3 unsigned
+  sVARIABLE+
```

Comment:
I think we need to add the number of variables here (but not to the bindings in generated ranges as they should always have the same number of bindings as the number of variables in the corresponding original scope).

Reply:
Whats the motivation? I skipped the number of variables on purpose and accept that we can't extend the `original_scope_variables` item in the future. Decoder can read VLQs until they hit the next `,`. And if we ever switch to a binary format, we'd prepend the length of the whole item either before or after the tag VLQ.

Reply:
> Decoder can read VLQs until they hit the next `,`

But in the grammar `original_scope_variables` is followed by `original_scope_source_index` or `original_scope_tree` or `original_scope_end`, without a `,` in between. We could add the `,` there instead of adding the number of variables.

Reply:
This was a mistake on my part. I updated the grammar to align more with the spec text grammar of https://szuend.github.io/ecma426-proposal-scopes/#sec-format.